### PR TITLE
Added Service Placement Interface and Intf Implementation

### PR DESCRIPTION
**Pull request by UtkarshBhatthere**
Status:
Mon: [X]
Mgr: [X]
Mds: [X]
Rgw: [X]

### CODE REVIEW

**Review by sabaini — CHANGES_REQUESTED** [review]
Thanks Utkarsh, did a quick review -- some inline commentary

File: microceph/api/services.go [hunk 1]
```diff
@@ -32,6 +32,30 @@ func cmdServicesGet(s *state.State, r *http.Request) response.Response {
 	return response.SyncResponse(true, services)
 }
 
+// Service Enable Endpoint.
+var enableServiceCmd = rest.Endpoint{
+	Path: "services/enable",
```
[the remark this answers is not in the document]
Comment (UtkarshBhatthere) [reply microceph/api/services.go]: Ack

File: microceph/ceph/bootstrap.go [hunk 2]
```diff
@@ -14,23 +14,23 @@ import (
 	"github.com/canonical/microceph/microceph/database"
 )
 
+// Common definition for bootstrap and service initialisation
+var confPath = filepath.Join(os.Getenv("SNAP_DATA"), "conf")
+var runPath = filepath.Join(os.Getenv("SNAP_DATA"), "run")
+var dataPath = filepath.Join(os.Getenv("SNAP_COMMON"), "data")
+var logPath = filepath.Join(os.Getenv("SNAP_COMMON"), "logs")
+
+var pathFileMode = map[string]os.FileMode{
```
[the remark this answers is not in the document]
Comment (UtkarshBhatthere) [reply microceph/ceph/bootstrap.go:23]: Ack

File: microceph/ceph/bootstrap.go [hunk 3]
```diff
@@ -14,23 +14,23 @@ import (
 	"github.com/canonical/microceph/microceph/database"
 )
 
+// Common definition for bootstrap and service initialisation
+var confPath = filepath.Join(os.Getenv("SNAP_DATA"), "conf")
+var runPath = filepath.Join(os.Getenv("SNAP_DATA"), "run")
+var dataPath = filepath.Join(os.Getenv("SNAP_COMMON"), "data")
+var logPath = filepath.Join(os.Getenv("SNAP_COMMON"), "logs")
```
[the remark this answers is not in the document]
Comment (UtkarshBhatthere) [reply microceph/ceph/bootstrap.go:21]: Ack

File: microceph/ceph/services_placement.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+import (
+	"fmt"
+
+	"github.com/canonical/microceph/microceph/api/types"
+	"github.com/canonical/microcluster/state"
+	"github.com/lxc/lxd/shared/logger"
+)
+
+type placementIntf interface {
+	PopulateParams(*state.State, string)(error)
```
[the remark this answers is not in the document]
Comment (UtkarshBhatthere) [reply microceph/ceph/services_placement.go:12]: Ack

File: microceph/ceph/services_placement.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+var servicePlacementTable = map[string](placementIntf) {
+	"mon": GenericServicePlacement{"mon"},
+	"mgr": GenericServicePlacement{"mgr"},
+	"mds": GenericServicePlacement{"mds"},
+	"rgw": RgwServicePlacement{},
+}
+
+func EnableService(s *state.State, payload types.EnableService) error {
```
[the remark this answers is not in the document]
Comment (UtkarshBhatthere) [reply microceph/ceph/services_placement.go:26]: Ack

File: microceph/ceph/services_placement_generic.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+	Name string
+}
+
+func (gsp GenericServicePlacement) PopulateParams(s *state.State, payload string) error {
+	// No params needed to initialise generic service
+	return nil
+}
+
+func (gsp GenericServicePlacement) HospitalityCheck(s *state.State) error {
+	return genericHospitalityCheck(gsp.Name)
```
[the remark this answers is not in the document]
Comment (UtkarshBhatthere) [reply microceph/ceph/services_placement_generic.go:33]: Yes, having this wrapper approach provides any developer implementing a new ServicePlacement to pick and choose these functions (RGWImplementation does) even while not implementing the GenericPlacement.

File: microceph/ceph/services_placement.go [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+type placementIntf interface {
+	PopulateParams(*state.State, string)(error)
+	HospitalityCheck(*state.State)(error)
+	ServiceInit(*state.State)(error)
+	PostPlacementCheck(*state.State)(error)
+	DbUpdate(*state.State)(error)
+}
+
+var servicePlacementTable = map[string](placementIntf) {
```
[the remark this answers is not in the document]
Comment (UtkarshBhatthere) [reply microceph/ceph/services_placement.go]: Ack

File: microceph/ceph/services_placement_generic.go [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+	"fmt"
+	"time"
+
+	"github.com/canonical/microceph/microceph/database"
+	"github.com/canonical/microcluster/state"
+	"github.com/lxc/lxd/shared/logger"
+)
+
+// Maps the addService function to respective services.
+var addServiceTable = map[string](func (string)(error)) {
```
[the remark this answers is not in the document]
Comment (UtkarshBhatthere) [reply microceph/ceph/services_placement_generic.go]: ack

File: microceph/ceph/manager.go [hunk 9]
```diff
@@ -27,3 +30,28 @@ func bootstrapMgr(hostname string, path string) error {
...
+		return fmt.Errorf("Failed to join manager: %w", err)
+	}
+
+	err = snapStart("mgr", true)
+	if err != nil {
+		return fmt.Errorf("Failed to start manager: %w", err)
+	}
+
+	return nil
+}
```
Comment (lmlg) [inline microceph/ceph/manager.go:57]: This pattern of adding a service by creating a directory, joining and starting the snap seems to be repeated a couple of times. Would it be worthy to create a helper for this?

File: microceph/ceph/services_placement.go [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+	if payload.Wait {
+		err := EnableService(s, payload, sp)
+		if err != nil {
+			logger.Errorf("Failed %s service enablement request: %v", payload.Name, err)
+			return err
+		}
+	} else {
+		go func() {
+			// Async call to Enable service.
+			EnableService(s, payload, sp)
```
Comment (lmlg) [inline microceph/ceph/services_placement.go:49]: what about error reporting here? Even if it's asynchronous, shouldn't we fetch the error and log it at least?

Comment (UtkarshBhatthere) [reply microceph/ceph/services_placement.go]: Added outer error log for EnableService in async case. EnableService itself has error logs at each failure.

File: microceph/cmd/microceph/enable_mds.go [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+func (c *cmdEnableMDS) Command() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "mds [--target <server>]",
+		Short: "Enable the MDS service on the --target server (default: this server)",
+		RunE:  c.Run,
+	}
+	cmd.PersistentFlags().StringVar(&c.flagTarget, "target", "", "Server hostname (default: this server)")
+	cmd.Flags().BoolVar(&c.wait, "wait", false, "Wait for mds service to be up.")
```
Comment (sabaini) [inline microceph/cmd/microceph/enable_mds.go:25]: Wondering if we shouldn't make `--wait=true` the default here. 

If I understand correctly, with `--wait=false` users won't see any errors unless they look in the log -- which isn't something a user would do routinely.

Otoh, any errors that might occur are directly reported back when `--wait=true` which feels like the safer and more explicit option. 

Same for the commands below

**Review by sabaini — CHANGES_REQUESTED** [review]
Hey, this is great work.

I've left some nitty detail requests inline mostly about style

File: microceph/ceph/monitor.go [hunk 12]
```diff
@@ -57,20 +57,20 @@ func bootstrapMon(hostname string, path string, monmap string, keyring string) e
 func joinMon(hostname string, path string) error {
 	tmpPath, err := os.MkdirTemp("", "")
 	if err != nil {
-		return fmt.Errorf("Unable to create temporary path: %w", err)
+		return fmt.Errorf("unable to create temporary path: %w", err)
```
[the remark this answers is not in the document]
Comment (UtkarshBhatthere) [reply microceph/ceph/monitor.go:60]: static check reports `error strings should not be capitalized (ST1005)go-staticcheck`

File: microceph/api/types/services.go [hunk 13]
```diff
@@ -10,6 +10,13 @@ type Service struct {
 	Location string `json:"location" yaml:"location"`
 }
 
+type EnableService struct {
```
[the remark this answers is not in the document]
Comment (UtkarshBhatthere) [reply microceph/api/types/services.go:13]: Ack

File: microceph/ceph/bootstrap.go [hunk 14]
```diff
@@ -116,7 +105,7 @@ func Bootstrap(s common.StateInterface) error {
 	return nil
 }
 
-func createKeyrings(confPath string) (string, error) {
+func createKeyrings(ConfPath string) (string, error) {
```
[the remark this answers is not in the document]
Comment (UtkarshBhatthere) [reply microceph/ceph/bootstrap.go:108]: Ack, fixed for all the methods.

File: microceph/ceph/rgw.go [hunk 15]
```diff
@@ -166,10 +164,10 @@ func createRGWKeyring(path string) error {
 }
 
 // symlinkRGWKeyring creates a symlink to the RGW keyring in the conf directory for use with the radosgw-admin command.
-func symlinkRGWKeyring(keyPath, confPath string) error {
+func symlinkRGWKeyring(keyPath, ConfPath string) error {
```
[the remark this answers is not in the document]
Comment (UtkarshBhatthere) [reply microceph/ceph/rgw.go:167]: Ack

File: microceph/ceph/services_placement_generic.go [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+	// Check in a loop if the mon service stays up.
+	attempts := 4
+
+	for attempts > 0 {
+		ret := snapCheckActive(service)
+		if ret != nil {
+			return ret
+		}
+
+		time.Sleep(time.Duration(attempts) * time.Second)
```
[the remark this answers is not in the document]
Comment (UtkarshBhatthere) [reply microceph/ceph/services_placement_generic.go:109]: ack

File: microceph/ceph/services_placement_test.go [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+func (s *servicesPlacementSuite) TestUnknownServiceFailure() {
+	payload := types.EnableService{
+		Name:    "unknowService",
+		Wait:    true,
+		Payload: "",
+	}
+
+	// Check Enable Service fails for unregistered services.
+	err := ServicePlacementHandler(s.TestStateInterface, payload)
+	assert.NotEqual(s.T(), err, nil)
```
[the remark this answers is not in the document]
Comment (UtkarshBhatthere) [reply microceph/ceph/services_placement_test.go:63]: Fixed for all tests.

File: microceph/ceph/services_placement_test.go [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+
+	payload := types.EnableService{
+		Name:    service,
+		Wait:    true,
+		Payload: "\"Port\":80", // Json String does not have {}
+	}
+
+	// Check Enable Service fails for unregistered services.
+	err := ServicePlacementHandler(s.TestStateInterface, payload)
+	assert.Contains(s.T(), err.Error(), "failed to populate the payload")
```
[the remark this answers is not in the document]
Comment (UtkarshBhatthere) [reply microceph/ceph/services_placement_test.go:77]: ack

File: microceph/cmd/microceph/enable_mds.go (hunk 11 above)
Comment (UtkarshBhatthere) [reply microceph/cmd/microceph/enable_mds.go:25]: Hmm, that does convince me to use wait=true for such instances.

File: microceph/cmd/microceph/enable_mds.go [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+
+type cmdEnableMDS struct {
+	common     *CmdControl
+	wait       bool
+	flagTarget string
+}
+
+func (c *cmdEnableMDS) Command() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "mds [--target <server>]",
```
[the remark this answers is not in the document]
Comment (UtkarshBhatthere) [reply microceph/cmd/microceph/enable_mds.go:20]: ack

File: microceph/ceph/rgw.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (sabaini) [reply microceph/ceph/rgw.go:166]: FAOD I'd suggest to revert to `confPath` here to make it clear this is a private var

File: microceph/ceph/services_placement.go [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+
+import (
+	"fmt"
+
+	"github.com/canonical/lxd/shared/logger"
+	"github.com/canonical/microceph/microceph/api/types"
+	"github.com/canonical/microceph/microceph/common"
+)
+
+type PlacementIntf interface {
```
[the remark this answers is not in the document]
Comment (sabaini) [reply microceph/ceph/services_placement.go:11]: Maybe something like

```
// PlacementIntf is an interface for placement managers which manage enablement of Ceph 
// services on a node
```

File: microceph/cmd/microceph/disable_rgw.go [hunk 21]
```diff
@@ -43,7 +43,7 @@ func (c *cmdDisableRGW) Run(cmd *cobra.Command, args []string) error {
 		Enabled: false,
 	}
 
-	err = client.EnableRGW(context.Background(), cli, req)
+	err = client.DisbleRGW(context.Background(), cli, req)
```
Comment (sabaini) [inline microceph/cmd/microceph/disable_rgw.go:46]: Typo: DisableRGW

File: microceph/client/services.go [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+	err := c.Query(queryCtx, "GET", api.NewURL().Path("services"), nil, &services)
+	if err != nil {
+		return nil, fmt.Errorf("Failed listing services: %w", err)
+	}
+
+	return services, nil
+}
+
+// DisbleRGW requests Ceph configures the RGW service.
+func DisbleRGW(ctx context.Context, c *client.Client, data *types.RGWService) error {
```
Comment (sabaini) [inline microceph/client/services.go:32]: Typo: `DisableRGW()`

File: microceph/ceph/services_placement_generic.go [hunk 23]
```diff
@@ -0,0 +1,130 @@
...
+func (gsp *GenericServicePlacement) DbUpdate(s common.StateInterface) error {
+	return genericDbUpdate(s, gsp.Name)
+}
+
+// Generic Method Implementations
+func genericHospitalityCheck(service string) error {
+	// Check if mon service already exists on host.
+	err := snapCheckActive(service)
+	if err == nil {
+		ret_err := fmt.Errorf("%s service already active on host", service)
```
[the remark this answers is not in the document]
Comment (UtkarshBhatthere) [reply microceph/ceph/services_placement_generic.go:91]: ack

File: microceph/ceph/services_placement_generic.go [hunk 24]
```diff
@@ -0,0 +1,130 @@
...
+	var ok bool
+	var addService func(string, string) error
+	hostname := s.ClusterState().Name()
+	pathConsts := common.GetPathConst()
+	pathFileMode := common.GetPathFileMode()
+	serviceDataPath := filepath.Join(pathConsts.DataPath, gsp.Name, fmt.Sprintf("ceph-%s", hostname))
+	addServiceTable := GetAddServiceTable()
+
+	// Fetch addService handler for gsp.Name service
+	if addService, ok = addServiceTable[gsp.Name]; !ok {
```
[the remark this answers is not in the document]
Comment (UtkarshBhatthere) [reply microceph/ceph/services_placement_generic.go:50]: ack

File: microceph/ceph/services_placement.go [hunk 25]
```diff
@@ -0,0 +1,102 @@
...
+	}
+}
+
+func ServicePlacementHandler(s common.StateInterface, payload types.EnableService) error {
+	var ok bool
+	var spt = GetServicePlacementTable()
+	var sp PlacementIntf
+
+	logger.Debugf("Enabling %s service, payload: %v", payload.Name, payload.Payload)
+	if sp, ok = spt[payload.Name]; !ok {
```
[the remark this answers is not in the document]
Comment (UtkarshBhatthere) [reply microceph/ceph/services_placement.go:40]: ack

**Comment by sabaini** [comment]
lgtm, cheers!

**Pull request merged**